File: src/main/java/code/回溯/Nqueen.java
package code.回溯;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Nqueen {
    
    List<List<String>> res = new ArrayList<>();
    public List<List<String>> solveNQueens(int n) {
        
        char[][] chessboard = new char[n][n];

        for (char[] ch : chessboard) {
            Arrays.fill(ch, '.');
        }

        backTracking(n, 0, chessboard);

        return res;
    }

    private void backTracking(int n, int row, char[][] chessboard) {
        if (row == n) {
            res.add(Array2List(chessboard));
            return;
        }

        for (int col = 0; col < chessboard.length; col++) {
            if (isValid(row, col, chessboard)) {
                chessboard[row][col] = 'Q';
                backTracking(n, row + 1, chessboard);
                chessboard[row][col] = '.';
            }
        }
    }

    private boolean isValid(int row, int col, char[][] chessboard) {

        // 判断列
        for (int i = 0; i < row; i++) {
            if (chessboard[i][col] == 'Q') return false;
        }

        // 判断45度角
        int i = row - 1, j = col - 1;
        for ( ; i >=0 && j >= 0; j--, i--) {
            if (chessboard[i][j] == 'Q') return false;
        }

        // 判断135度角
        for ( i = row - 1, j = col + 1 ; i >=0 && j < chessboard.length; j++, i--) {
            if (chessboard[i][j] == 'Q') return false;
        }
        return true;
    }

    private List<String> Array2List(char[][] chessboard) {
        List<String> list = new ArrayList<>();
        for (char[] ch : chessboard) {
            list.add(String.valueOf(ch));
        }
        return list;
    }
}

File: src/main/python/NQueen.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        # res 存放最终结果
        # chessboard 存放每次的结果

        # n * n 棋盘，深度就是 行 n，树的宽度也是 n，树的宽度也是n
        # 满足 row == n return 
        # 每次遍历 col 时都需要判断 chessbord是否有效 

        res = []
        chessboard = ['.' * n for _ in range(n)] # 初始化棋盘

        self.backTrancking(n, 0, chessboard, res)

        return res

    def backTrancking(self, n, row, chessboard, res):
        if row == n:
            res.append(chessboard[:])
            return

        for col in range(n):
            if self.isValid(row, col, chessboard):
                chessboard[row] = chessboard[row][:col] + 'Q' + chessboard[row][col+1:]
                self.backTrancking(n, row + 1, chessboard, res)
                chessboard[row] = chessboard[row][:col] + '.' + chessboard[row][col+1:]

    def isValid(self, row, col, chessboard):
        # 检查列
        for i in range(row):
            if chessboard[i][col] == 'Q':
                return False

        # 检查45度角
        i, j = row - 1, col - 1
        while i >= 0 and j >= 0:
            if chessboard[i][j] == 'Q':
                return False
            i -= 1
            j -= 1
        
        # 检查135度角
        i, j = row - 1, col + 1
        while i >= 0 and j < len(chessboard):
            if chessboard[i][j] == 'Q':
                return False
            i -= 1
            j += 1
        return True

        